src/Cryptol/TypeCheck/Infer.hs

```diff
@@ -1019,54 +1035,202 @@ checkSigB b (Forall as asmps0 t0, validSchema) = case thing (P.bDef b) of
+          -- constraint disjunctions, we encode a disjunction of conjunctions as
+          -- `[[Prop]]`. For exhaustive checking, each disjunct will need to be
+          -- checked independently, and all must pass in order to be considered
+          -- exhaustive.
+
+          let negateSimpleNumProp :: Prop -> Maybe [Prop]
+              negateSimpleNumProp prop = case prop of
+                TCon tcon tys -> case tcon of
+                  PC pc -> case pc of
+                    -- not x == y  <=>  x /= y
+                    PEqual -> pure [TCon (PC PNeq) tys]
+                    -- not x /= y  <=>  x == y
+                    PNeq -> pure [TCon (PC PEqual) tys]
+                    -- not x >= y  <=>  x /= y and y >= x
+                    PGeq -> pure [TCon (PC PNeq) tys, TCon (PC PGeq) (reverse tys)]
+                    -- not fin x  <=>  x == Inf
+                    PFin | [ty] <- tys -> pure [TCon (PC PEqual) [ty, tInf]]
+                         | otherwise -> panicInvalid
+                    -- not True  <=>  0 == 1
+                    PTrue -> pure [TCon (PC PEqual) [tZero, tOne]]
+                    _ -> mempty
+                  TC _tc -> panicInvalid
+                  TF _tf -> panicInvalid
+                  TError _ki -> panicInvalid -- TODO: where does this come from?
+                TUser _na _tys ty -> negateSimpleNumProp ty
+                _ -> panicInvalid
+                where
+                  panicInvalid = tcPanic "checkSigB"
+                    [ "This type shouldn't be a valid type constraint: " ++
+                      "`" ++ pretty prop ++ "`"]
+
+              negateSimpleNumProps :: [Prop] -> [[Prop]]
+              negateSimpleNumProps props = do
+                prop <- props
+                maybe mempty pure (negateSimpleNumProp prop)
+
+              toGoal :: Prop -> Goal
+              toGoal prop =
+                Goal
+                  { goalSource = CtPropGuardsExhaustive name
+                  , goalRange = srcRange $ P.bName b
+                  , goal = prop }
+
+              canProve :: [Prop] -> [Goal] -> InferM Bool
+              canProve asmps goals = isRight <$>
+                tryProveImplication (Just name) as asmps goals
+
+              -- Try to prove that the first guard will be satisfied. If cannot,
+              -- then assume it is false (via `negateSimpleNumProps`) and try to
+              -- prove that the second guard will be satisfied. If cannot, then
+              -- assume it is false, and so on. If the last guard cannot be
+              -- proven in this way, then issue a `NonExhaustivePropGuards`
+              -- warning. Note that when assuming that a conjunction of multiple
+              -- constraints is false, this results in a disjunction, and so
+              -- must check exhaustive under assumption that each disjunct is
+              -- false independently. 
+              -- 
```

Comment:
This is the derivation I came up with.  For a declaration:
```
AS =>

  | P, Q => ..
  | R, S => ..
  | T, U => ..
```
We'd like to show that:
```
AS => (P && Q) || (R && S) || (T && U)
```
This is equivalent to:
```
AS && !(P&&Q) && !(R&&S) => T && U
AS && (!P || !Q) && (!R || !S) => T && U
```
Now you have to consider how each assumed disjunction could be true, in this example this requires 4 proofs:
```
   AS && !P && !R => T && U
   AS && !P && !S => T && U
   AS && !Q && !R => T && U
   AS && !Q && !S => T && U
```
In the first step we have some freedom to choose which of the disjuncts (i.e., guards) to move to the LHS.   It is best to move the ones with fewest conjunctions in the guard, as this would result in the fewest proof obligations.

Reply:
Hmm, that is a more efficient way of checking than what I implemented, which is formulated as the following:
```
A => (P /\ Q) \/ (R /\ S) \/ (T /\ U)

<=>

(
  (A => P /\ Q)
) 
\/
( 
  (A /\ -P => R /\ S) \/ 
  (A /\ -Q => R /\ S) 
) 
\/
(
  (A /\ -P /\ -R => T /\ U) /\ 
  (A /\ -P /\ -S => T /\ U) /\ 
  (A /\ -Q /\ -R => T /\ U) /\
  (A /\ -Q /\ -S => T /\ U)
)
\/
(
  (A /\ -P /\ -R /\ -T => True) /\ 
  (A /\ -P /\ -R /\ -U => True) /\ 
  (A /\ -P /\ -S /\ -T => True) /\ 
  (A /\ -P /\ -S /\ -U => True) /\ 
  (A /\ -Q /\ -R /\ -T => True) /\
  (A /\ -Q /\ -R /\ -U => True) /\
  (A /\ -Q /\ -S /\ -T => True) /\
  (A /\ -Q /\ -S /\ -U => True)
)
```

Reply:
I'll try implementing it the more efficient way, as you formulated